### ISSUE TITLE
error of TypeError: unhashable type: 'ConfigProto'  when running an example code at tensorflow extended

### ISSUE THREAD

**Issue by umusa** [tensorflow/transform#188]
I got an error when I run the code at https://www.tensorflow.org/tfx/tutorials/tfx/components_keras

on databricks notebook.

      TypeError: unhashable type: 'ConfigProto' [while running 'Analyze/PackedCombineAccumulate[ApplySavedModel[Phase0]]/InitialPackedCombineGlobally/CombinePerKey/WindowIntoDiscarding']

       databricks env:  7.0 ML Spark 3.0.0 Scala 2.12, GPU
       python: 3.7.6
       tensorflow: 2.2.0
       tensorflow extended: 0.22.0
     
Because the module created by the python magic command "%%wirtefile" cannot be found and imported, I made some changes to the code.

The new code: (wrapped the cell contents into a class)
        
   ```
class Taxi_constants():
  
         # Categorical features are assumed to each have a maximum value in the dataset.
  MAX_CATEGORICAL_FEATURE_VALUES = [24, 31, 12]

  CATEGORICAL_FEATURE_KEYS = [
      'trip_start_hour', 'trip_start_day', 'trip_start_month',
      'pickup_census_tract', 'dropoff_census_tract', 'pickup_community_area',
      'dropoff_community_area'
  ]
```

  ```
DENSE_FLOAT_FEATURE_KEYS = ['trip_miles', 'fare', 'trip_seconds']

  # Number of buckets used by tf.transform for encoding each feature.
  FEATURE_BUCKET_COUNT = 10

  BUCKET_FEATURE_KEYS = [
      'pickup_latitude', 'pickup_longitude', 'dropoff_latitude',
      'dropoff_longitude'
  ]

  # Number of vocabulary terms used for encoding VOCAB_FEATURES by tf.transform
  VOCAB_SIZE = 1000

  # Count of out-of-vocab buckets in which unrecognized VOCAB_FEATURES are hashed.
  OOV_SIZE = 10

  VOCAB_FEATURE_KEYS = [
      'payment_type',
      'company',
  ]

  # Keys
  LABEL_KEY = 'tips'
  FARE_KEY = 'fare'

  def transformed_name(key):
    return key + '_xf'
   
```
       _taxi_transform_module_file = 'taxi_transform.py'

```

%%writefile {_taxi_transform_module_file}

import tensorflow as tf
import tensorflow_transform as tft

#import taxi_constants.        # because this import does not work, a class (Taxi_constants) is created here.
                                                # not sure why this module cannot be imported even though it has been written to 
                                                # a file by "%%write_to_file"
#%%writefile {_taxi_constants_module_file}
class Taxi_constants():
  
  def __init__(self):
    # Categorical features are assumed to each have a maximum value in the dataset.
    self.MAX_CATEGORICAL_FEATURE_VALUES = [24, 31, 12]

    self.CATEGORICAL_FEATURE_KEYS = [
        'trip_start_hour', 'trip_start_day', 'trip_start_month',
        'pickup_census_tract', 'dropoff_census_tract', 'pickup_community_area',
        'dropoff_community_area'
    ]

    self.DENSE_FLOAT_FEATURE_KEYS = ['trip_miles', 'fare', 'trip_seconds']

    # Number of buckets used by tf.transform for encoding each feature.
    self.FEATURE_BUCKET_COUNT = 10

    self.BUCKET_FEATURE_KEYS = [
        'pickup_latitude', 'pickup_longitude', 'dropoff_latitude',
        'dropoff_longitude'
    ]

    # Number of vocabulary terms used for encoding VOCAB_FEATURES by tf.transform
    self.VOCAB_SIZE = 1000

    # Count of out-of-vocab buckets in which unrecognized VOCAB_FEATURES are hashed.
    self.OOV_SIZE = 10

    self.VOCAB_FEATURE_KEYS = [
        'payment_type',
        'company',
    ]

    # Keys
    self.LABEL_KEY = 'tips'
    self.FARE_KEY = 'fare'

  def transformed_name(self, key):
    return key + '_xf'
  
taxi_constants = Taxi_constants()

_DENSE_FLOAT_FEATURE_KEYS = taxi_constants.DENSE_FLOAT_FEATURE_KEYS
_VOCAB_FEATURE_KEYS = taxi_constants.VOCAB_FEATURE_KEYS
_VOCAB_SIZE = taxi_constants.VOCAB_SIZE
_OOV_SIZE = taxi_constants.OOV_SIZE
_FEATURE_BUCKET_COUNT = taxi_constants.FEATURE_BUCKET_COUNT
_BUCKET_FEATURE_KEYS = taxi_constants.BUCKET_FEATURE_KEYS
_CATEGORICAL_FEATURE_KEYS = taxi_constants.CATEGORICAL_FEATURE_KEYS
_FARE_KEY = taxi_constants.FARE_KEY
_LABEL_KEY = taxi_constants.LABEL_KEY
_transformed_name = taxi_constants.transformed_name


def preprocessing_fn(inputs):
  """tf.transform's callback function for preprocessing inputs.
  Args:
    inputs: map from feature keys to raw not-yet-transformed features.
  Returns:
    Map from string feature key to transformed feature operations.
  """
  outputs = {}
  for key in _DENSE_FLOAT_FEATURE_KEYS:
    # Preserve this feature as a dense float, setting nan's to the mean.
    #print(key)
    outputs[_transformed_name(key)] = tft.scale_to_z_score(
        _fill_in_missing(inputs[key]))

  for key in _VOCAB_FEATURE_KEYS:
    # Build a vocabulary for this feature.
    #print(key)
    outputs[_transformed_name(key)] = tft.compute_and_apply_vocabulary(
        _fill_in_missing(inputs[key]),
        top_k=_VOCAB_SIZE,
        num_oov_buckets=_OOV_SIZE)

  for key in _BUCKET_FEATURE_KEYS:
    #print(key)
    outputs[_transformed_name(key)] = tft.bucketize(
        _fill_in_missing(inputs[key]), _FEATURE_BUCKET_COUNT)

  for key in _CATEGORICAL_FEATURE_KEYS:
    #print(key)
    outputs[_transformed_name(key)] = _fill_in_missing(inputs[key])

  # Was this passenger a big tipper?
  #print(_FARE_KEY)
  taxi_fare = _fill_in_missing(inputs[_FARE_KEY])
  tips = _fill_in_missing(inputs[_LABEL_KEY])
  outputs[_transformed_name(_LABEL_KEY)] = tf.where(
      tf.math.is_nan(taxi_fare),
      tf.cast(tf.zeros_like(taxi_fare), tf.int64),
      # Test if the tip was > 20% of the fare.
      tf.cast(
          tf.greater(tips, tf.multiply(taxi_fare, tf.constant(0.2))), tf.int64))

  return outputs


def _fill_in_missing(x):
  """Replace missing values in a SparseTensor.
  Fills in missing values of `x` with '' or 0, and converts to a dense tensor.
  Args:
    x: A `SparseTensor` of rank 2.  Its dense shape should have size at most 1
      in the second dimension.
  Returns:
    A rank 1 tensor where missing values of `x` have been filled in.
  """
  default_value = '' if x.dtype == tf.string else 0
  return tf.squeeze(
      tf.sparse.to_dense(
          tf.SparseTensor(x.indices, x.values, [x.dense_shape[0], 1]),
          default_value),
      axis=1)

```
  
output:

        Overwriting taxi_transform.py

```
transform = Transform(
    examples=example_gen.outputs['examples'],
    schema=schema_gen.outputs['schema'],
    module_file=os.path.abspath(_taxi_transform_module_file))
context.run(transform)

```

output:

        INFO:tensorflow:Saver not created because there are no variables in the graph to restore
        TypeError: unhashable type: 'ConfigProto' [while running 'Analyze/PackedCombineAccumulate[ApplySavedModel[Phase0]]/InitialPackedCombineGlobally/CombinePerKey/WindowIntoDiscarding']

**Comment by zoyahav** [comment]
What version of tensorflow-transform are you using?
And can you please provide a full stacktrace?

**Comment by umusa** [comment]
I posted the stacktrace, which is a little bit long.

TypeError                                 Traceback (most recent call last)
/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker.invoke_process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker._invoke_process_per_window()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common._OutputProcessor.process_outputs()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.PGBKCVOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.PGBKCVOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/combiners.py in create_accumulator(self)
    696   def create_accumulator(self):
--> 697     return [c.create_accumulator() for c in self._combiners]
    698 

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/combiners.py in <listcomp>(.0)
    696   def create_accumulator(self):
--> 697     return [c.create_accumulator() for c in self._combiners]
    698 

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/beam/analyzer_impls.py in create_accumulator(self)
    680   def create_accumulator(self):
--> 681     return self._combiner.create_accumulator()
    682 

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/analyzers.py in create_accumulator(self)
   1547   def create_accumulator(self):
-> 1548     graph_state = self._get_graph_state()
   1549     return graph_state.empty_summary

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/analyzers.py in _get_graph_state(self)
   1544     return _global_quantiles_graph_state_provider.get_graph_state(
-> 1545         self._graph_state_options)
   1546 

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/analyzers.py in get_graph_state(self, graph_state_options)
   1869     # of some objects that might otherwise be avoided).
-> 1870     result = self._graph_states_by_options.get(graph_state_options)
   1871     if result is None:

TypeError: unhashable type: 'ConfigProto'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<command-2472572> in <module>
      3     schema=schema_gen.outputs['schema'],
      4     module_file=os.path.abspath(_taxi_transform_module_file))
----> 5 context.run(transform)

/databricks/python/lib/python3.7/site-packages/tfx/orchestration/experimental/interactive/interactive_context.py in run_if_ipython(*args, **kwargs)
     64       # __IPYTHON__ variable is set by IPython, see
     65       # https://ipython.org/ipython-doc/rel-0.10.2/html/interactive/reference.html#embedding-ipython.
---> 66       return fn(*args, **kwargs)
     67     else:
     68       absl.logging.warning(

/databricks/python/lib/python3.7/site-packages/tfx/orchestration/experimental/interactive/interactive_context.py in run(self, component, enable_cache, beam_pipeline_args)
    166         component, pipeline_info, driver_args, metadata_connection,
    167         beam_pipeline_args, additional_pipeline_args)
--> 168     execution_id = launcher.launch().execution_id
    169 
    170     return execution_result.ExecutionResult(

/databricks/python/lib/python3.7/site-packages/tfx/orchestration/launcher/base_component_launcher.py in launch(self)
    203                          execution_decision.input_dict,
    204                          execution_decision.output_dict,
--> 205                          execution_decision.exec_properties)
    206 
    207     absl.logging.info('Running publisher for %s',

/databricks/python/lib/python3.7/site-packages/tfx/orchestration/launcher/in_process_component_launcher.py in _run_executor(self, execution_id, input_dict, output_dict, exec_properties)
     65         executor_context)  # type: ignore
     66 
---> 67     executor.Do(input_dict, output_dict, exec_properties)

/databricks/python/lib/python3.7/site-packages/tfx/components/transform/executor.py in Do(self, input_dict, output_dict, exec_properties)
    389       label_outputs[labels.CACHE_OUTPUT_PATH_LABEL] = cache_output
    390     status_file = 'status_file'  # Unused
--> 391     self.Transform(label_inputs, label_outputs, status_file)
    392     absl.logging.debug('Cleaning up temp path %s on executor success',
    393                        temp_path)

/databricks/python/lib/python3.7/site-packages/tfx/components/transform/executor.py in Transform(***failed resolving arguments***)
    950                       compute_statistics,
    951                       per_set_stats_output_paths,
--> 952                       materialization_format)
    953   # TODO(b/122478841): Writes status to status file.
    954 

/databricks/python/lib/python3.7/site-packages/tfx/components/transform/executor.py in _RunBeamImpl(self, use_tfxio, analyze_data_list, transform_data_list, preprocessing_fn, input_dataset_metadata, transform_output_path, raw_examples_data_format, temp_path, input_cache_dir, output_cache_dir, compute_statistics, per_set_stats_output_paths, materialization_format)
   1325                | 'Materialize[{}]'.format(infix) >> self._WriteExamples(
   1326                    materialization_format,
-> 1327                    dataset.materialize_output_path))
   1328 
   1329     return _Status.OK()

/databricks/python/lib/python3.7/site-packages/apache_beam/pipeline.py in __exit__(self, exc_type, exc_val, exc_tb)
    545     try:
    546       if not exc_type:
--> 547         self.run().wait_until_finish()
    548     finally:
    549       self._extra_context.__exit__(exc_type, exc_val, exc_tb)

/databricks/python/lib/python3.7/site-packages/apache_beam/pipeline.py in run(self, test_runner_api)
    524         finally:
    525           shutil.rmtree(tmpdir)
--> 526       return self.runner.run_pipeline(self, self._options)
    527     finally:
    528       shutil.rmtree(self.local_tempdir, ignore_errors=True)

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in run_pipeline(self, pipeline, options)
    171 
    172     self._latest_run_result = self.run_via_runner_api(
--> 173         pipeline.to_runner_api(default_environment=self._default_environment))
    174     return self._latest_run_result
    175 

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in run_via_runner_api(self, pipeline_proto)
    181     # TODO(pabloem, BEAM-7514): Create a watermark manager (that has access to
    182     #   the teststream (if any), and all the stages).
--> 183     return self.run_stages(stage_context, stages)
    184 
    185   @contextlib.contextmanager

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in run_stages(self, stage_context, stages)
    329           stage_results = self._run_stage(
    330               runner_execution_context,
--> 331               bundle_context_manager,
    332           )
    333           monitoring_infos_by_stage[stage.name] = (

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in _run_stage(self, runner_execution_context, bundle_context_manager)
    506               input_timers,
    507               expected_timer_output,
--> 508               bundle_manager)
    509 
    510       final_result = merge_results(last_result)

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in _run_bundle(self, runner_execution_context, bundle_context_manager, data_input, data_output, input_timers, expected_timer_output, bundle_manager)
    544 
    545     result, splits = bundle_manager.process_bundle(
--> 546         data_input, data_output, input_timers, expected_timer_output)
    547     # Now we collect all the deferred inputs remaining from bundle execution.
    548     # Deferred inputs can be:

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in process_bundle(self, inputs, expected_outputs, fired_timers, expected_output_timers, dry_run)
    928     with UnboundedThreadPoolExecutor() as executor:
    929       for result, split_result in executor.map(execute, zip(part_inputs,  # pylint: disable=zip-builtin-not-iterating
--> 930                                                             timer_inputs)):
    931         split_result_list += split_result
    932         if merged_result is None:

/databricks/python/lib/python3.7/concurrent/futures/_base.py in result_iterator()
    596                     # Careful not to keep a reference to the popped future
    597                     if timeout is None:
--> 598                         yield fs.pop().result()
    599                     else:
    600                         yield fs.pop().result(end_time - time.monotonic())

/databricks/python/lib/python3.7/concurrent/futures/_base.py in result(self, timeout)
    433                 raise CancelledError()
    434             elif self._state == FINISHED:
--> 435                 return self.__get_result()
    436             else:
    437                 raise TimeoutError()

/databricks/python/lib/python3.7/concurrent/futures/_base.py in __get_result(self)
    382     def __get_result(self):
    383         if self._exception:
--> 384             raise self._exception
    385         else:
    386             return self._result

/databricks/python/lib/python3.7/site-packages/apache_beam/utils/thread_pool_executor.py in run(self)
     42       # If the future wasn't cancelled, then attempt to execute it.
     43       try:
---> 44         self._future.set_result(self._fn(*self._fn_args, **self._fn_kwargs))
     45       except BaseException as exc:
     46         # Even though Python 2 futures library has #set_exection(),

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in execute(part_map_input_timers)
    924           input_timers,
    925           expected_output_timers,
--> 926           dry_run)
    927 
    928     with UnboundedThreadPoolExecutor() as executor:

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/fn_runner.py in process_bundle(self, inputs, expected_outputs, fired_timers, expected_output_timers, dry_run)
    824             process_bundle_descriptor.id,
    825             cache_tokens=[next(self._cache_token_generator)]))
--> 826     result_future = self._worker_handler.control_conn.push(process_bundle_req)
    827 
    828     split_results = []  # type: List[beam_fn_api_pb2.ProcessBundleSplitResponse]

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/portability/fn_api_runner/worker_handlers.py in push(self, request)
    351       self._uid_counter += 1
    352       request.instruction_id = 'control_%s' % self._uid_counter
--> 353     response = self.worker.do_instruction(request)
    354     return ControlFuture(request.instruction_id, response)
    355 

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/sdk_worker.py in do_instruction(self, request)
    469       # E.g. if register is set, this will call self.register(request.register))
    470       return getattr(self, request_type)(
--> 471           getattr(request, request_type), request.instruction_id)
    472     else:
    473       raise NotImplementedError

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/sdk_worker.py in process_bundle(self, request, instruction_id)
    504         with self.maybe_profile(instruction_id):
    505           delayed_applications, requests_finalization = (
--> 506               bundle_processor.process_bundle(instruction_id))
    507           monitoring_infos = bundle_processor.monitoring_infos()
    508           monitoring_infos.extend(self.state_cache_metrics_fn())

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/bundle_processor.py in process_bundle(self, instruction_id)
    970           elif isinstance(element, beam_fn_api_pb2.Elements.Data):
    971             input_op_by_transform_id[element.transform_id].process_encoded(
--> 972                 element.data)
    973 
    974       # Finish all operations.

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/bundle_processor.py in process_encoded(self, encoded_windowed_values)
    216       decoded_value = self.windowed_coder_impl.decode_from_stream(
    217           input_stream, True)
--> 218       self.output(decoded_value)
    219 
    220   def monitoring_infos(self, transform_id, tag_to_pcollection_id):

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.Operation.output()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.Operation.output()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner._reraise_augmented()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.SimpleInvoker.invoke_process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common._OutputProcessor.process_outputs()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner._reraise_augmented()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker.invoke_process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker._invoke_process_per_window()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common._OutputProcessor.process_outputs()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.ConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner._reraise_augmented()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.SimpleInvoker.invoke_process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common._OutputProcessor.process_outputs()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner._reraise_augmented()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.SimpleInvoker.invoke_process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common._OutputProcessor.process_outputs()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.DoOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner._reraise_augmented()

/databricks/python/lib/python3.7/site-packages/future/utils/__init__.py in raise_with_traceback(exc, traceback)
    444         if traceback == Ellipsis:
    445             _, _, traceback = sys.exc_info()
--> 446         raise exc.with_traceback(traceback)
    447 
    448 else:

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.DoFnRunner.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker.invoke_process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common.PerWindowInvoker._invoke_process_per_window()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/common.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.common._OutputProcessor.process_outputs()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.SingletonConsumerSet.receive()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.PGBKCVOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/runners/worker/operations.cpython-37m-x86_64-linux-gnu.so in apache_beam.runners.worker.operations.PGBKCVOperation.process()

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/combiners.py in create_accumulator(self)
    695 
    696   def create_accumulator(self):
--> 697     return [c.create_accumulator() for c in self._combiners]
    698 
    699   def merge_accumulators(self, accumulators):

/databricks/python/lib/python3.7/site-packages/apache_beam/transforms/combiners.py in <listcomp>(.0)
    695 
    696   def create_accumulator(self):
--> 697     return [c.create_accumulator() for c in self._combiners]
    698 
    699   def merge_accumulators(self, accumulators):

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/beam/analyzer_impls.py in create_accumulator(self)
    679 
    680   def create_accumulator(self):
--> 681     return self._combiner.create_accumulator()
    682 
    683   def add_input(self, accumulator, next_input):

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/analyzers.py in create_accumulator(self)
   1546 
   1547   def create_accumulator(self):
-> 1548     graph_state = self._get_graph_state()
   1549     return graph_state.empty_summary
   1550 

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/analyzers.py in _get_graph_state(self)
   1543   def _get_graph_state(self):
   1544     return _global_quantiles_graph_state_provider.get_graph_state(
-> 1545         self._graph_state_options)
   1546 
   1547   def create_accumulator(self):

/databricks/python/lib/python3.7/site-packages/tensorflow_transform/analyzers.py in get_graph_state(self, graph_state_options)
   1868     # population is thread-safe (even if it might occasionally waste creation
   1869     # of some objects that might otherwise be avoided).
-> 1870     result = self._graph_states_by_options.get(graph_state_options)
   1871     if result is None:
   1872       result = _QuantilesGraphState(graph_state_options)

TypeError: unhashable type: 'ConfigProto' [while running 'Analyze/PackedCombineAccumulate[ApplySavedModel[Phase0]]/InitialPackedCombineGlobally/CombinePerKey/WindowIntoDiscarding']

**Comment by umusa** [comment]
Can you please let me know how to find transform version ? I do not find it on https://www.tensorflow.org/tfx/api_docs/python/tfx/components/transform

**Comment by zoyahav** [comment]
You should be able to just run:
`!pip show tensorflow-transform`

**Comment by umusa** [comment]
Name: tensorflow-transform
Version: 0.22.0
Summary: A library for data preprocessing with TensorFlow
Home-page: https://www.tensorflow.org/tfx/transform
Author: Google Inc.
Author-email: tensorflow-extended-dev@googlegroups.com
License: Apache 2.0
Location: /databricks/conda/envs/databricks-ml-gpu/lib/python3.7/site-packages
Requires: pydot, tensorflow-metadata, numpy, six, absl-py, apache-beam, tfx-bsl, tensorflow, protobuf
Required-by: tfx, tensorflow-data-validation

**Comment by zoyahav** [comment]
Thanks, could you also please verify the version you have installed of protobuf?
`!pip show protobuf`

**Comment by umusa** [comment]
Name: protobuf
Version: 3.11.4
Summary: Protocol Buffers
Home-page: https://developers.google.com/protocol-buffers/
Author: None
Author-email: None
License: 3-Clause BSD License
Location: /databricks/conda/envs/databricks-ml-gpu/lib/python3.7/site-packages
Requires: six, setuptools
Required-by: tfx, tfx-bsl, tensorflow, tensorflow-transform, tensorflow-serving-api, tensorflow-model-analysis, tensorflow-metadata, tensorflow-gpu, tensorflow-data-validation, tensorboard, mlflow, ml-metadata, googleapis-common-protos, google-cloud-bigquery, google-api-core, apache-beam

**Comment by umusa** [comment]
Could anybody help me about this ? thanks!

**Comment by iindyk** [comment]
We're actively working on this issue at the moment

**Comment by umusa** [comment]
Thanks for your response. Is it possible to get a temporary work-around for this ? 
We depend on tensorflow-extended to work on some ML things. :)

**Comment by iindyk** [comment]
Nothing simple comes to mind, but the error is not observed for runners different from Spark Runner. So you may try to change the environment if possible

**Comment by umusa** [comment]
We have to work on databricks Spark, which is the only one environment. This issue seems to be version problem because "Configproto" is deprecated in TF 2.0 and beyond ?

**Comment by umusa** [comment]
Does TF and TFX can only be run in Jupyter on local machine or google cloud ?

**Comment by iindyk** [comment]
It's not deprecated as of now and no, it should be able to run in your environment as well. That's an issue on our side

**Comment by umusa** [comment]
If TFX work well on google cloud (with spark backend) but does not on databricks environment (also with spark), there could be some different settings in the both environment such that TFX got this issue on databricks ?  
Or TFX does not work on all Spark environments ? 
thanks

**Comment by umusa** [comment]
We run TFX on AWS/EC2 through databricks.

**Comment by iindyk** [comment]
[The issue](https://issues.apache.org/jira/browse/SPARK-22674) seems to be unrelated to TFT -- Spark affects the behavior of subclasses of `collections.namedtuple` after deserialization. Here's a simple code to showcase the issue:
```python
from collections import namedtuple
import pickle
# import pyspark


class MyNamedTuple(namedtuple('MyNamedTuple', ['hello'])):
    def __str__(self):
        return 'Hello, this is MyNamedTuple'


result = MyNamedTuple('hello')
result = pickle.loads(pickle.dumps(result))

print(result)
print(isinstance(result, MyNamedTuple))
```
It's output is as expected:
```
Hello, this is MyNamedTuple
True
```
However, if pyspark import is uncommented:
```
MyNamedTuple(hello='hello')
False
```

**Comment by iindyk** [comment]
The TypeError you're seeing is caused by the fact that custom `__hash__` function for [_QuantilesGraphStateOptions](https://github.com/tensorflow/transform/blob/a43404e3470359f4ed0d777b4960838633a3238e/tensorflow_transform/analyzers.py#L2068) is not used since it's a subclass of `collections.namedtuple`.

If you want to use TFT with Spark, there is a slightly painful workaround to this until the Spark issue is resolved: you can locally modify subclasses of `collections.namedtuple` that you encounter to something equivalent. Number of these changes, however, depends on your code. But since you're using `bucketize` (or `quantiles`), you'll need to modify at least [_QuantilesGraphStateOptions](https://github.com/tensorflow/transform/blob/a43404e3470359f4ed0d777b4960838633a3238e/tensorflow_transform/analyzers.py#L2068) and [_TensorBinding](https://github.com/tensorflow/transform/blob/a43404e3470359f4ed0d777b4960838633a3238e/tensorflow_transform/beam/impl.py#L469).

**Comment by iindyk** [comment]
Fixed by https://github.com/tensorflow/transform/commit/d17afa3a12b038be342d6630e9556d010712cb23
